gh-117657: Use an atomic store to set type flags.

nascheme opened the pull request:
The `PyType_HasFeature()` function reads the flags with a relaxed atomic load and without holding the type lock.  To avoid data races, use atomic stores to update the type flags.

This change eliminates some warnings from TSAN.  For non-free-threaded builds, there should be no performance impact.

<!-- gh-issue-number: gh-117657 -->
* Issue: gh-117657
<!-- /gh-issue-number -->

nascheme committed:
Use an atomic store to set type flags.

The `PyType_HasFeature()` function reads the flags with a relaxed atomic
load and without holding the type lock.  To avoid data races, use atomic
stores if `PyType_Ready()` has already been called.

ZeroIntensity commented on Objects/typeobject.c:
```diff
@@ -343,6 +344,49 @@ _PyStaticType_GetBuiltins(void)
+type_set_flags(PyTypeObject *tp, unsigned long flags)
+{
+#ifdef Py_GIL_DISABLED
+    if (tp->tp_flags & Py_TPFLAGS_READY) {
```
Shouldn't the read operation be atomic as well?

nascheme replied:
I don't think so.  This is used either if the `TYPE_LOCK` is held or if the type is newly created and the lock will be held later, before exposing the type to other threads.  So there is no way for another thread to write to the flags.

colesbury commented on Objects/typeobject.c:
```diff
@@ -343,6 +344,49 @@ _PyStaticType_GetBuiltins(void)
+type_set_flags(PyTypeObject *tp, unsigned long flags)
+{
+#ifdef Py_GIL_DISABLED
+    if (tp->tp_flags & Py_TPFLAGS_READY) {
```
I think we should just use `FT_ATOMIC_STORE_ULONG_RELAXED`:

* The relaxed atomic store is essentially a normal store, other than the occasional slightly worse code gen we've chatted about.
* In general, with a few exceptions, I don't think these sorts of optimizations are worth the added complexity even if this were an expensive atomic operation.
* In the default build, `FT_ATOMIC_STORE_ULONG_RELAXED(tp->tp_flags, flags)` expands to `tp->tp_flags = flags`

nascheme replied:
Okay, seems reasonable to me.

colesbury commented on Objects/typeobject.c:
```diff
@@ -343,6 +344,49 @@ _PyStaticType_GetBuiltins(void)
+static void
+type_add_flags(PyTypeObject *tp, unsigned long flag)
+{
+    type_set_flags(tp, tp->tp_flags | flag);
```
`type_add_flags` and `type_clear_flags` do not have an `ASSERT_TYPE_LOCK_HELD()`. Are they called without the type lock?

nascheme replied:
Yes, they are sometimes called without the lock.  That should be done only in places were the type is being initially created and the lock will be held later, before the type is revealed.  At least, that's how I think it needs to work.

I would either try to fix that code so it takes the lock or I could make the assert something like `if <type ready> then ASSERT_TYPE_LOCK_HELD()`

colesbury reviewed (commented):
Overall, looks good to me. I think `type_set_flags` can be simplified and I have a question below about locking.

nascheme committed:
Simplify `type_set_flags()` function.

Just use FT_ATOMIC_STORE_ULONG_RELAXED() always, not worth the extra
code complexity.

nascheme committed:
Add type lock held assert to `type_set_flags()`.

This helps ensure that type flag setting is done in a free-threading
safe way. Fix a few cases were flags were set without holding the type
lock. Use `type_modified_unlocked()` in a couple places where the lock
is already head.

nascheme committed:
Merge 'origin/main' into gh-117657-tp-flags-set-atomic
